### PR TITLE
HWY-293: Request latest state only if stalled.

### PR DESCRIPTION
**Pull request by goral09**
With the `TIMER_ID_PANORAMA_REQUEST` firing periodically, a node would send out (to every peer) a request containing its latest panorama. The receiver, would then compare it with its local view of the DAG and either request a dependency from the sender, or send a newer unit (that the sender doesn't have yet). It would do that for every validator. This leads to `m * m * n` requests being made every `request_latest_panorama`, where `m` is the number of nodes in the network and `n` number of validators – everyone (`m`) would send a message to everyone else (`m - 1`) and trigger potentially `n` responses (one for every validator in the panorama), some of which (`RequestDependency`) would trigger a response (yet another message). That amount of additional messages being exchanged turned out to really hurt the performance of the network. We need to remove the periodic message.

Originally, the reason for adding it was to prevent the node from being left alone in an era, while others have already moved on to the next one. The alternative solution proposed here is that rather than sending `LatestStateRequest` periodically, we will do it only when the node detects it's stalled. If, after the timer fires again, the node's protocol state hasn't advanced yet – shut down.

https://casperlabs.atlassian.net/browse/HWY-293

### CODE REVIEW

**Review by EdHastingsCasperLabs — APPROVED** [review]
lgtm

File: node/src/components/consensus/protocols/highway.rs [hunk 1]
```diff
@@ -462,6 +461,33 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
...
+        if self.evidence_only || self.finalized_switch_block() {
+            return vec![]; // Era has ended. No further progress is expected.
+        }
+        if self.last_panorama == *self.highway.state().panorama() {
+            // We haven't made any progress. Request latest panorama from peers and schedule
+            // standstill alert. If we still won't progress by the time
+            // `TIMER_ID_STANDSTILL_ALERT` is handled, it means we're stuck.
+            let mut outcomes = self.latest_panorama_request();
+            outcomes.push(ProtocolOutcome::ScheduleTimer(
+                now + self.standstill_timeout * 2,
```
Comment (fizyk20) [inline node/src/components/consensus/protocols/highway.rs:477]: I'm not sure about this `*2` here. I mean, we did discuss making the wait after the initial check longer, but this is a recurring timer that will be firing every `self.standstill_timeout`, anyway, so I'm not sure if it makes much sense to make only this single period longer.
Not a blocker, though.

**Review by fizyk20 — APPROVED** [review]
Nice one :+1:

File: node/src/components/consensus/protocols/highway.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/consensus/protocols/highway.rs:477]: Yeah, I wasn't convinced either (didn't have a strong opinion about it) but wanted to take your suggestion into account. Will revert it.

Comment (goral09) [reply node/src/components/consensus/protocols/highway.rs]: Removed in [`f3b1a5c` (#1397)](https://github.com/casper-network/casper-node/pull/1397/commits/f3b1a5c25b031cd6c04328b2e84d863a18c5a2d1)

**Comment by goral09** [comment]
bors merge

File: node/src/components/consensus/protocols/highway.rs [hunk 2]
```diff
@@ -457,6 +457,33 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
...
+    // If we haven't, we might have been left alone in the era and we should request the state from
+    // peers.
+    fn handle_progress_alert_timer(&mut self, now: Timestamp) -> ProtocolOutcomes<I, C> {
+        if self.evidence_only || self.finalized_switch_block() {
+            return vec![]; // Era has ended. No further progress is expected.
+        }
+        if self.last_panorama == *self.highway.state().panorama() {
+            // We haven't made any progress. Request latest panorama from peers and schedule
+            // standstill alert. If we still won't progress by the time
+            // `TIMER_ID_STANDSTILL_ALERT` is handled, it means we're stuck.
```
Comment (afck) [inline node/src/components/consensus/protocols/highway.rs:470]: `TIMER_ID_STANDSTILL_ALERT` → `TIMER_ID_PROGRESS_ALERT`

**Pull request merged**